### ISSUE TITLE
File Extension Agnostic Schema Inference 

### ISSUE THREAD

**Issue by tustvold** [apache/arrow-datafusion#1736]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I wrote a test approximating

```
let file = tempfile::tempfile();

// ... write parquet data ...

let mut context = ExecutionContext::new();
context.register_parquet("t", file.path().as_str())
context.sql("select column from t");
```

This would result in "Invalid identifier" errors, effectively claiming the column didn't exist. I verified the file existed, had the correct columns, etc... I was very confused :laughing: 

Eventually I tracked this down to the schema being inferred as empty if the extension is not ".parquet", this feels unexpected

**Describe the solution you'd like**

Either "register_parquet" should return an error if the extension is missing, or `FileFormat::infer_schema` should be more agnostic to file extensions.

**Comment by alamb** [comment]
Change this description to be a bug, which I think better reflects what is going on

**Comment by Ted-Jiang** [comment]
@tustvold  may i ask you how to write a tmp parquet file, IMHP i think all parquet data file should end .parquet.

**Comment by tustvold** [comment]
@Ted-Jiang Remove this line in the SQL benchmark https://github.com/apache/arrow-datafusion/pull/1738/files#diff-d1dbff8af63c3a3fe4d918432f982181b40fa4b7e1641522a6a48904f521fc89R143 and that was what caused issues.

I'm don't feel strongly whether the `.parquet` file should be mandatory or not, but I do think silently inferring an empty schema makes for a pretty poor UX :sweat_smile:

**Comment by alamb** [comment]
>  but I do think silently inferring an empty schema makes for a pretty poor UX 😅

I agree -- an error about "can not infer schema" seems much better than silently ignoring

**Comment by andygrove** [comment]
I plan on fixing this for 14.0.0

**Comment by andygrove** [comment]
Some notes from investigating this.

This is where the `EmptyExec` gets introduced. I ran into this issue in Ballista as well in https://github.com/apache/arrow-ballista/pull/414

```rust
        // if no files need to be read, return an `EmptyExec`
        if partitioned_file_lists.is_empty() {
            let schema = self.schema();
            let projected_schema = project_schema(&schema, projection.as_ref())?;
            return Ok(Arc::new(EmptyExec::new(false, projected_schema)));
        }
```

The problem is I am registering a CSV data source and the filename does not end with `.csv` and we have hard-coded assumptions that CSV files must end with `".csv"`, so we cannot support `.tsv` or `.tbl`. I think we just need to add an error check for this case and prompt the user to specify the file extension they want rather than use the default.

**Comment by andygrove** [comment]
I propose that the error check we add is that at least one file exists with the specified extension.

**Comment by alamb** [comment]
> I propose that the error check we add is that at least one file exists with the specified extension.



Makes sense to me

**Comment by liningpan** [comment]
I also ran into this issue. The problem seems to be even if only a single file is provided, we still try to match by extension. 

https://github.com/apache/arrow-datafusion/blob/0e5f6df2c4fb2d647874102a19c74eaaf7f34d98/datafusion/core/src/datasource/listing/url.rs#L143-L175

If my use case always uses a single csv file for each table, would a reasonable workaround be setting file extension to an empty string in `CsvReadOptions`? Are there any problems?

```rust
ctx.register_csv(
    name,
    path,
    CsvReadOptions::new().file_extension("")
).await?
```

**Comment by andygrove** [comment]
Maybe `file_extension` could be made an `Option<String>`, defaulting to `None`?
